include/CL/sycl/buffer/detail/buffer.hpp

```diff
@@ -68,15 +82,23 @@ class buffer : public detail::buffer_base,
   */
   boost::multi_array_ref<T, Dimensions> access;
 
-  /// The weak pointer to copy back data on buffer deletion
-  weak_ptr_class<T> final_data;
+  /* How to copy back data on buffer destruction, can be modified with
+   * set_final_data( ... )
+   */
+  boost::optional<std::function<void(void)>> final_write_back;
+
+  // Used to store the shared pointer used as argument of set_final_data
```

Comment:
No this is a normal `shared_ptr` and it is used just to keep alive the memory of the buffer at construction time.